File: src/services/node.service.ts
import type { Node } from '@/types';
import { SEED_NODES } from '@/mock/seed';

const delay = (min = 200, max = 1000): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve, Math.random() * (max - min) + min));

let nodes = [...SEED_NODES];

export const nodeService = {
  async getAll(): Promise<Node[]> {
    await delay(300, 800);
    return [...nodes];
  },

  async toggleFavorite(id: string): Promise<Node> {
    await delay(100, 300);
    const index = nodes.findIndex((n) => n.id === id);
    if (index === -1) throw new Error(`Node ${id} not found`);
    nodes[index] = { ...nodes[index]!, isFavorite: !nodes[index]!.isFavorite };
    return { ...nodes[index]! };
  },

  async testDelay(id: string): Promise<{ delay: number; loss: number }> {
    await delay(500, 2000);
    const delay_ = Math.floor(Math.random() * 350) + 5;
    const loss = Math.random() * 5;
    const index = nodes.findIndex((n) => n.id === id);
    if (index !== -1) {
      nodes[index] = { ...nodes[index]!, delay: delay_, loss, status: 'online' };
    }
    return { delay: delay_, loss };
  },

  async testAllDelay(): Promise<void> {
    await delay(1000, 5000);
    nodes = nodes.map((node) => ({
      ...node,
      delay: Math.floor(Math.random() * 350) + 5,
      loss: Math.random() * 5,
      status: Math.random() > 0.1 ? ('online' as const) : ('offline' as const),
    }));
  },

  async connect(id: string): Promise<Node> {
    await delay(200, 600);
    nodes = nodes.map((n) => ({
      ...n,
      isConnected: n.id === id,
    }));
    const node = nodes.find((n) => n.id === id);
    if (!node) throw new Error(`Node ${id} not found`);
    return { ...node, isConnected: true };
  },

  async disconnect(id: string): Promise<Node> {
    await delay(100, 300);
    const index = nodes.findIndex((n) => n.id === id);
    if (index === -1) throw new Error(`Node ${id} not found`);
    nodes[index] = { ...nodes[index]!, isConnected: false };
    return { ...nodes[index]! };
  },
};

File: src-tauri/src/service/node_service.rs
use rand::Rng;

use crate::core::AppState;
use crate::models::{Node, NodeStatus};
use crate::utils::AppResult;

pub fn get_all(state: &AppState) -> Vec<Node> {
    state.nodes.read().clone()
}

pub fn toggle_favorite(state: &AppState, id: &str) -> AppResult<Node> {
    let mut nodes = state.nodes.write();
    let node = nodes
        .iter_mut()
        .find(|n| n.id == id)
        .ok_or_else(|| crate::utils::AppError::NotFound(format!("Node {}", id)))?;
    node.is_favorite = !node.is_favorite;
    Ok(node.clone())
}

/// Result of a delay test.
#[derive(Debug, Clone, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct NodeDelayResult {
    pub delay: i64,
    pub loss: f64,
}

/// Test delay for a single node. Returns random latency values.
pub fn test_delay(state: &AppState, id: &str) -> AppResult<NodeDelayResult> {
    let mut rng = rand::thread_rng();
    let delay: i64 = rng.gen_range(5..350);
    let loss: f64 = rng.gen_range(0.0..5.0);

    // Update node in place
    let mut nodes = state.nodes.write();
    if let Some(node) = nodes.iter_mut().find(|n| n.id == id) {
        node.delay = Some(delay);
        node.loss = Some(loss);
        node.status = if rng.gen::<f64>() > 0.1 {
            NodeStatus::Online
        } else {
            NodeStatus::Offline
        };
    }

    Ok(NodeDelayResult { delay, loss })
}

/// Test delay for all nodes.
pub fn test_all_delay(state: &AppState) {
    let mut rng = rand::thread_rng();
    let mut nodes = state.nodes.write();

    for node in nodes.iter_mut() {
        node.delay = Some(rng.gen_range(5..350));
        node.loss = Some(rng.gen_range(0.0..5.0));
        node.status = if rng.gen::<f64>() > 0.1 {
            NodeStatus::Online
        } else {
            NodeStatus::Offline
        };
    }
}

/// Connect to a node (mock — just sets is_connected).
pub fn connect(state: &AppState, id: &str) -> AppResult<Node> {
    let mut nodes = state.nodes.write();

    // Disconnect all others first
    for node in nodes.iter_mut() {
        node.is_connected = false;
    }

    let target = nodes
        .iter_mut()
        .find(|n| n.id == id)
        .ok_or_else(|| crate::utils::AppError::NotFound(format!("Node {}", id)))?;

    target.is_connected = true;
    target.status = NodeStatus::Online;

    // Enable system proxy on connect (via ResourceManager)
    if let Some(rm) = state.get_resource_manager() {
        let port = state.get_config().mixed_port;
        let _ = rm.platform_manager.enable_system_proxy("127.0.0.1", port);
        rm.event_bus.publish(crate::event::AppEvent::StatusChange(
            crate::models::DashboardRunStatus::Running,
        ));
    }

    Ok(target.clone())
}

/// Disconnect from a node (mock).
pub fn disconnect(state: &AppState, id: &str) -> AppResult<Node> {
    let mut nodes = state.nodes.write();
    let node = nodes
        .iter_mut()
        .find(|n| n.id == id)
        .ok_or_else(|| crate::utils::AppError::NotFound(format!("Node {}", id)))?;

    node.is_connected = false;

    // Disable system proxy on disconnect
    if let Some(rm) = state.get_resource_manager() {
        let _ = rm.platform_manager.disable_system_proxy();
        rm.event_bus.publish(crate::event::AppEvent::StatusChange(
            crate::models::DashboardRunStatus::Stopped,
        ));
    }

    Ok(node.clone())
}
